### ISSUE TITLE
Synchronize all pending operations

### ISSUE THREAD

**Issue by dionhaefner** [google/jax#4335]
With `mpi4jax` we are running into a problem when Python exits before a pending operation has the chance to be executed.

When using MPI, `MPI_Finalize` runs at interpreter exit. Any MPI operation called *after* that will deadlock, which can happen due to JAX's  asynchronous dispatch.

**Is there a way to force synchronization of *all* pending XLA operations?** Then we could register that as an `atexit` handler that runs before `MPI_Finalize`.

`block_until_ready` doesn't work for 2 reasons:

1. Additional boilerplate in user code.
2. Some of our primitive only return an XLA token which doesn't have a `block_until_ready` method.

**Comment by jekbradbury** [comment]
On all current XLA platforms, there is only one (default) compute stream per device. That means any asynchronous operation will be enqueued at the _end_ of the current queue for that compute stream, meaning that `jax.device_put(0.).block_until_ready()` will be a valid way to synchronize the whole device. (We could also add a way to do this directly, but that should work for now!)

**Comment by dionhaefner** [comment]
That sounds promising, but unfortunately it doesn't seem to work. I can add the line you proposed to the end of the script and see that it runs *before* my custom XLA call. Is this a CPU vs GPU thing? We're only running on CPU so far.

It does seem to work if I JIT it, though:

```python
jax.jit(lambda: jax.device_put)(0.).block_until_ready()
```

Does that make sense? It's hard to tell whether this fixes the problem or it just works because it takes longer so execute (so the race condition is less likely to be triggered).

**Comment by jekbradbury** [comment]
You’re right, of course! The snippet I gave doesn’t synchronize the compute stream because it runs on the independent memcpy stream. Jitting it as in your snippet (or running anything that launches an XLA computation) will synchronize the compute stream.

**Comment by dionhaefner** [comment]
I use this now:

```python
(jax.device_put(0.) + 0).block_until_ready()
```

which works OK.

An official way to do this would be nice, though, as this workaround seems a bit brittle (relying on JAX implementation details).

**Comment by dionhaefner** [comment]
Either way, thanks for the help, of course :)
